<a href="https://colab.research.google.com/github/Trabit/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/Copy_of_LS_DS12_112_Make_Features.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 2*

---

# Make Features 

- Student should be able to understand the purpose of feature engineering
- Student should be able to work with strings in pandas
- Student should be able to work with dates and times in pandas
- Student should be able to modify or create columns of a dataframe using the `.apply()` function


Helpful Links:
- [Minimally Sufficient Pandas](https://medium.com/dunder-data/minimally-sufficient-pandas-a8e67f2a2428)
- [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)
- Python Data Science Handbook
  - [Chapter 3.10](https://jakevdp.github.io/PythonDataScienceHandbook/03.10-working-with-strings.html), Vectorized String Operations
  - [Chapter 3.11](https://jakevdp.github.io/PythonDataScienceHandbook/03.11-working-with-time-series.html), Working with Time Series
- [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)

# [Objective](#feature-engineering) - The Purpose of Feature Engineering



## Overview

Feature Engineering is the process of using a combination of domain knowledge, creativity and the pre-existing columns of a dataset to create completely new columns.

 Machine Learning models try to detect patterns in the data and then associate those patterns with certain predictions. The hope is that by creating new columns on our dataset that we can expose our model to new patterns in the data so that it can make better and better predictions.

This is largely a matter of understanding how to work with individual columns of a dataframe with Pandas --which is what we'll be practicing today!

## Follow Along

Columns of a dataframe hold each hold a specific type of data. Lets inspect some of the common datatypes found in datasets and then we'll make a new feature on a dataset using pre-existing columns.

In [0]:
import pandas as pd

In [8]:
# Lets take a look at the Ames Iowa Housing Dataset:

import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd            

### Specific Columns hold specific kinds of data

In [7]:
#pandas disply options
import pandas as pd

pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)

df.dtypes

Id                 int64
MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
Alley             object
LotShape          object
LandContour       object
Utilities         object
LotConfig         object
LandSlope         object
Neighborhood      object
Condition1        object
Condition2        object
BldgType          object
HouseStyle        object
OverallQual        int64
OverallCond        int64
YearBuilt          int64
YearRemodAdd       int64
RoofStyle         object
RoofMatl          object
Exterior1st       object
Exterior2nd       object
MasVnrType        object
MasVnrArea       float64
ExterQual         object
ExterCond         object
Foundation        object
BsmtQual          object
BsmtCond          object
BsmtExposure      object
BsmtFinType1      object
BsmtFinSF1         int64
BsmtFinType2      object
BsmtFinSF2         int64
BsmtUnfSF          int64
TotalBsmtSF        int64
Heating           object


Some columns hold integer values like the `BedroomAbvGr` which stands for "Bedrooms Above Grade." This is the number of non-basement bedrooms in the home.

For more information on specific column meanings view the [data dictionary](https://github.com/ryanleeallred/datasets/blob/master/Ames%20Housing%20Data/data_description.txt).

In [10]:
# Look at the first ten rows of the `BedroomAbvGr` column.
# Looks like integers to me!
df['BedroomAbvGr'].head()

0    3
1    3
2    3
3    3
4    4
Name: BedroomAbvGr, dtype: int64

Some columns hold float values like the `LotFrontage` column.

In [13]:
# Look at the first ten rows of the 'LotFrontage` column.
df['LotFrontage'].head(10)


0    65.0
1    80.0
2    68.0
3    60.0
4    84.0
5    85.0
6    75.0
7     NaN
8    51.0
9    50.0
Name: LotFrontage, dtype: float64

Hmmm, do the values above look like floats to you?

They all have .0 on them so technically they're being stored as floats, but *should* they be stored as floats?

Lets see what all of the possible values for this column are.

In [14]:
df['LotFrontage'].value_counts(dropna=False)

NaN      259
60.0     143
70.0      70
80.0      69
50.0      57
        ... 
106.0      1
38.0       1
138.0      1
140.0      1
137.0      1
Name: LotFrontage, Length: 111, dtype: int64

Looks to me like the `LotFrontage` column originally held integer values but was cast to a `float` meaning that each original integer values was converted to its corresponding float representation. 

Any guesses as to why that would have happened?


HINT: What's the most common `LotFrontage` value for this column?

In [15]:
# NaN is the most common value in this column. What is a NaN
df['LotFrontage']. value_counts(dropna=False).head()

NaN     259
60.0    143
70.0     70
80.0     69
50.0     57
Name: LotFrontage, dtype: int64

`NaN` stands stands for "Not a Number" and is the default missing value indicator with Pandas. This means there were cells in this column that didn't have a LotFrontage value recorded for those homes. 

This is where domain knowledge starts to come in. Think about the context we're working with here: houses. What might a null or blank cell representing "Linear feet of street connected to property" mean in the context of a housing dataset?

Ok, so maybe it makes seanse to have some NaNs in this column. What is the datatype of a NaN value?

Perhaps some of this data is truly missing or unrecorded data, but sometimes `NaNs` are more likely to indicate something that was "NA" or "Not Applicable" to a particular observation. There could be multiple reasons why there was no value recorded for a particular feature.

Remember, that Pandas tries to maintain a single datatype for all values in a column, and therefore...

In [16]:
import numpy as np

# What is the datatype of NaN?

my_var = 10.0

type(np.NaN)

float

The datatype of a NaN is float!  This means that if we have a column of integer values, but the column has even a single `NaN` that column will not be treated with the integer datatype but all of the integers will be converted to floats in order to try and preserve the same datatype throughout the entire column.

You can see already how understanding column datatypes is crucial to understanding how Pandas help us manage our data.

### Making new Features

Lets slim down the dataset and consider just a few specific columns:

- `TotalBsmtSF`
- `1stFlrSF`
- `2ndFlrSF`
- `SalePrice1`


In [18]:
# I can make a smaller dataframe with a few specific column headers
# by passing a list of column headers inside of the square brackets
small_df = df[['TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'SalePrice']]

small_df

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice
0             856       856       854     208500
1            1262      1262         0     181500
2             920       920       866     223500
3             756       961       756     140000
4            1145      1145      1053     250000
...           ...       ...       ...        ...
1455          953       953       694     175000
1456         1542      2073         0     210000
1457         1152      1188      1152     266500
1458         1078      1078         0     142125
1459         1256      1256         0     147500

[1460 rows x 4 columns]

### Syntax for creating new columns

When making a new column on a dataframe, we have to use the square bracket syntax of accessing a column. We can't use "dot syntax" here.

In [23]:
# Lets add up all of the square footage to get a single square footage 
# column for the entire dataset

#Dot Syntax - I CANNOT use dot sytax when creating
df.SalePrice

# Bracket Syntax - I have to use Bracket Syntax
df['SalePrice']

# Using bracket syntax to make a new 'TotalSquareFootage' cloumn
small_df['TotalSquareFootage'] = small_df['TotalBsmtSF'] + small_df['1stFlrSF'] + small_df['2ndFlrSF']

small_df

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage
0             856       856       854     208500                2566
1            1262      1262         0     181500                2524
2             920       920       866     223500                2706
3             756       961       756     140000                2473
4            1145      1145      1053     250000                3343
...           ...       ...       ...        ...                 ...
1455          953       953       694     175000                2600
1456         1542      2073         0     210000                3615
1457         1152      1188      1152     266500                3492
1458         1078      1078         0     142125                2156
1459         1256      1256         0     147500                2512

[1460 rows x 5 columns]

In [25]:
# Lets make a nother new column that is 'PricePerSqFt' by
# dividing the price by the square footage

small_df['PricePerSqft'] = small_df['SalePrice'] / small_df['TotalSquareFootage']

small_df.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSquareFootage  \
0          856       856       854     208500                2566   
1         1262      1262         0     181500                2524   
2          920       920       866     223500                2706   
3          756       961       756     140000                2473   
4         1145      1145      1053     250000                3343   

   PricePerSqft  
0     81.254871  
1     71.909667  
2     82.594235  
3     56.611403  
4     74.783129

Ok, we have made two new columns on our small dataset.

- What does a **high** `PricePerSqFt` say about a home that the square footage and price alone don't capture as directly?

- What does a **low** `PricePerSqFt` say about a home that the square footage and price alone don't directly capture?



## Challenge

I hope you can see how we have used existing columns to create a new column on a dataset that say something new about our unit of observation. This is what making new features (columns) on a dataset is all about and why it's so essential to data science --particularly predictive modeling "Machine Learning." 

We'll spend the rest of the lecture and assignment today trying to get as good as we can at manipulating (cleaning) and creating new columns on datasets.

# [Objective](#work-with-strings) Work with Strings with Pandas

## Overview

So far we have worked with numeric datatypes (ints and floats) but we haven't worked with any columns containing string values. We can't simply use arithmetic to manipulate string values, so we'll need to learn some more techniques in order to work with this datatype.

## Follow Along

We're going to import a new dataset here to work with. This dataset is from LendingClub and holds information about loans issued in Q4 of 2018. This dataset is a bit messy so it will give us plenty of opportunities to clean up existing columns as well as create new ones.

The `!wget` shell command being used here does exactly the same thing that your browser does when you type a URL in the address. It makes a request or "gets" the file at that address. However, in our case the file isn't a webpage, it's a compressed CSV file. 

Try copying and pasting the URL from below into your browser, did it start an automatic download? Any URLs like this that start automatic downloads when navigated to can be used along with the `!wget` command to bring files directly into your notebook's memory.

### Load a new dataset

In [26]:
!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip

--2020-01-18 00:49:23--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 64.48.1.20
Connecting to resources.lendingclub.com (resources.lendingclub.com)|64.48.1.20|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [               <=>  ]  21.72M   438KB/s    in 53s     

2020-01-18 00:50:16 (422 KB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [22776791]



We need to use the `!unzip` command to extract the csv from the zipped folder.

In [29]:
!unzip LoanStats_2018Q4.csv.zip

Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


We can also use bash/shell commands to look at the raw file using the `!head` and `!tail` commands

In [31]:
!head LoanStats_2018Q4.csv

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","op

In [32]:
!tail LoanStats_2018Q4.csv

"","","5600","5600","5600"," 36 months"," 13.56%","190.21","C","C1","","n/a","RENT","15600","Not Verified","Oct-2018","Current","n","","","credit_card","Credit card refinancing","836xx","ID","15.31","0","Aug-2012","0","","97","9","1","5996","34.5%","11","w","3686.37","3686.37","2652.39","2652.39","1913.63","738.76","0.0","0.0","0.0","Dec-2019","190.21","Jan-2020","Nov-2019","0","","1","Individual","","","","0","0","5996","0","0","0","1","20","0","","0","2","3017","35","17400","1","0","0","3","750","4689","45.5","0","0","20","73","13","13","0","13","","20","","0","3","5","4","4","1","9","10","5","9","0","0","0","0","100","25","1","0","17400","5996","8600","0","","","","","","","","","","","","N","","","","","","","","","","","","","","","N","","","","","",""
"","","23000","23000","23000"," 36 months"," 15.02%","797.53","C","C3","Tax Consultant","10+ years","MORTGAGE","75000","Source Verified","Oct-2018","Charged Off","n","","","debt_consolidation","Debt consolidation","352xx","AL","20.9

As we look at the raw file itself, do you see anything that might cause us trouble as we read in the CSV file to a dataframe?

In [34]:
# Read in the CSV

df = pd.read_csv('LoanStats_2018Q4.csv', header=1)
print(df.shape)
df.head()


/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(128399, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  NaN        NaN     6000.0       6000.0           6000.0   36 months   
1  NaN        NaN    24000.0      24000.0          24000.0   36 months   
2  NaN        NaN    30000.0      30000.0          30000.0   36 months   
3  NaN        NaN     8000.0       8000.0           8000.0   36 months   
4  NaN        NaN    22400.0      22400.0          22400.0   60 months   

  int_rate  installment grade sub_grade                emp_title emp_length  \
0   14.47%       206.44     C        C2                  Manager  10+ years   
1   10.33%       778.14     B        B1      Maintenance Manager  10+ years   
2   15.02%      1040.26     C        C3             Firefighter   10+ years   
3   12.98%       269.48     B        B5               Controller    4 years   
4   16.14%       546.40     C        C4  Paramedic/ Fire Fighter  10+ years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     55000.0            Verified  Dec-2018  Charged Off   
1       MORTGAGE    100000.0        Not Verified  Dec-2018      Current   
2           RENT     90000.0            Verified  Dec-2018   Fully Paid   
3       MORTGAGE     95000.0     Source Verified  Dec-2018      Current   
4            OWN     56000.0     Source Verified  Dec-2018      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    104xx   
1          n  NaN   NaN  debt_consolidation       Debt consolidation    606xx   
2          n  NaN   NaN  debt_consolidation       Debt consolidation    681xx   
3          n  NaN   NaN         credit_card  Credit card refinancing    480xx   
4          n  NaN   NaN         credit_card  Credit card refinancing    234xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  10.65          0.0         Jun-2006             0.0   
1         IL  19.14          0.0         Oct-1996             3.0   
2         NE  22.23          0.0         May-2007             1.0   
3         MI   9.44          0.0         Mar-2009             2.0   
4         VA  18.47          0.0         Dec-2004             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN       6.0      0.0   
1                    39.0                     NaN      14.0      0.0   
2                     NaN                     NaN       8.0      0.0   
3                    73.0                     NaN       7.0      0.0   
4                    75.0                     NaN      15.0      0.0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0     4196.0      91.2%        8.0                   w       0.00   
1     8547.0      27.7%       28.0                   w   16807.30   
2    25754.0        48%       23.0                   w       0.00   
3    11434.0      33.3%       20.0                   w    1707.34   
4    13555.0      38.2%       22.0                   w   19230.91   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   1106.520000          1106.52           546.14   
1       16807.30   9317.020000          9317.02          7192.70   
2           0.00  30883.673712         30883.67         30000.00   
3        1707.34   6855.630000          6855.63          6292.66   
4       19230.91   6526.670000          6526.67          3169.09   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         272.38                 0.0       288.0                    51.84   
1        2124.32                 0.0         0.0                     0.00   
2         883.67                 0.0         0.0                     0.00   
3         562.97                 0.0         0.0                     0.00   
4        3357.58                 0.0         0.0                     0.00   

  last_pymn

The extra rows at the top and bottom of the file have done two things:

1) The top row has made it so that the entire dataset is being interpreted as column headers

2) The bottom two rows have been read into the 'id' column and are causing every column to have at least two `NaN` values in it.

In [0]:
# We can fix the header problem by using the 'skiprows' parameter

Lets look at the NaN values of each column so that you can see the problem that the extra rows at the bottom of the file are creating for us

In [35]:
# Sum null values by column and sort from least to greatest
pd.set_option('display.max_rows',200)

df.isnull().sum().sort_values()

inq_fi                                             2
delinq_amnt                                        2
chargeoff_within_12_mths                           2
acc_open_past_24mths                               2
inq_last_12m                                       2
total_cu_tl                                        2
total_rev_hi_lim                                   2
max_bal_bc                                         2
open_rv_24m                                        2
open_rv_12m                                        2
total_bal_il                                       2
open_il_24m                                        2
hardship_flag                                      2
open_act_il                                        2
open_acc_6m                                        2
tot_cur_bal                                        2
tot_coll_amt                                       2
acc_now_delinq                                     2
application_type                              

In [37]:
# Address the extra NaNs in each column by skipping the footer as well.

df = pd.read_csv('LoanStats_2018Q4.csv', header=1, skipfooter=2, engine='python')
print(df.shape)
df.head()

(128397, 144)


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0 NaN        NaN       6000         6000           6000.0   36 months   
1 NaN        NaN      24000        24000          24000.0   36 months   
2 NaN        NaN      30000        30000          30000.0   36 months   
3 NaN        NaN       8000         8000           8000.0   36 months   
4 NaN        NaN      22400        22400          22400.0   60 months   

  int_rate  installment grade sub_grade                emp_title emp_length  \
0   14.47%       206.44     C        C2                  Manager  10+ years   
1   10.33%       778.14     B        B1      Maintenance Manager  10+ years   
2   15.02%      1040.26     C        C3             Firefighter   10+ years   
3   12.98%       269.48     B        B5               Controller    4 years   
4   16.14%       546.40     C        C4  Paramedic/ Fire Fighter  10+ years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     55000.0            Verified  Dec-2018  Charged Off   
1       MORTGAGE    100000.0        Not Verified  Dec-2018      Current   
2           RENT     90000.0            Verified  Dec-2018   Fully Paid   
3       MORTGAGE     95000.0     Source Verified  Dec-2018      Current   
4            OWN     56000.0     Source Verified  Dec-2018      Current   

  pymnt_plan  url  desc             purpose                    title zip_code  \
0          n  NaN   NaN         credit_card  Credit card refinancing    104xx   
1          n  NaN   NaN  debt_consolidation       Debt consolidation    606xx   
2          n  NaN   NaN  debt_consolidation       Debt consolidation    681xx   
3          n  NaN   NaN         credit_card  Credit card refinancing    480xx   
4          n  NaN   NaN         credit_card  Credit card refinancing    234xx   

  addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
0         NY  10.65            0         Jun-2006               0   
1         IL  19.14            0         Oct-1996               3   
2         NE  22.23            0         May-2007               1   
3         MI   9.44            0         Mar-2009               2   
4         VA  18.47            0         Dec-2004               0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN         6        0   
1                    39.0                     NaN        14        0   
2                     NaN                     NaN         8        0   
3                    73.0                     NaN         7        0   
4                    75.0                     NaN        15        0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0       4196      91.2%          8                   w       0.00   
1       8547      27.7%         28                   w   16807.30   
2      25754        48%         23                   w       0.00   
3      11434      33.3%         20                   w    1707.34   
4      13555      38.2%         22                   w   19230.91   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   1106.520000          1106.52           546.14   
1       16807.30   9317.020000          9317.02          7192.70   
2           0.00  30883.673712         30883.67         30000.00   
3        1707.34   6855.630000          6855.63          6292.66   
4       19230.91   6526.670000          6526.67          3169.09   

   total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
0         272.38                 0.0       288.0                    51.84   
1        2124.32                 0.0         0.0                     0.00   
2         883.67                 0.0         0.0                     0.00   
3         562.97                 0.0         0.0                     0.00   
4        3357.58                 0.0         0.0                     0.00   

  last_pymnt_d  

In [38]:
df.isnull().sum().sort_values

<bound method Series.sort_values of id                                            128397
member_id                                     128397
loan_amnt                                          0
funded_amnt                                        0
funded_amnt_inv                                    0
term                                               0
int_rate                                           0
installment                                        0
grade                                              0
sub_grade                                          0
emp_title                                      20942
emp_length                                     11701
home_ownership                                     0
annual_inc                                         0
verification_status                                0
issue_d                                            0
loan_status                                        0
pymnt_plan                                         0
url       

For good measure, we'll also drop some columns that are made up completely of NaN values.

Why might LendingClub have included columns in their dataset that are 100% blank?

In [39]:
df.shape

(128397, 144)

In [43]:
df = df.drop(['url', 'member_id', 'desc', 'id'], axis=1)

KeyError: ignored

In [44]:
df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3             Firefighter   10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     55000.0            Verified  Dec-2018  Charged Off          n   
1    100000.0        Not Verified  Dec-2018      Current          n   
2     90000.0            Verified  Dec-2018   Fully Paid          n   
3     95000.0     Source Verified  Dec-2018      Current          n   
4     56000.0     Source Verified  Dec-2018      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Jun-2006               0                     NaN   
1            0         Oct-1996               3                    39.0   
2            0         May-2007               1                     NaN   
3            0         Mar-2009               2                    73.0   
4            0         Dec-2004               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019           546

### Clean up the `int_rate` column

When we're preparing a dataset for a machine learning model we typically want to represent don't want to leave any string values in our dataset --because it's hard to do math on words. 

Specifically, we have a column that is representing a numeric value, but currently doesn't have a numeric datatype. Lets look at the first 10 values of the `int_rate` column

In [45]:
# Look at the first 10 values of the int_rate column
df['int_rate'].head(10)

0     14.47%
1     10.33%
2     15.02%
3     12.98%
4     16.14%
5     12.98%
6     15.02%
7     16.14%
8     14.47%
9     13.56%
Name: int_rate, dtype: object

In [46]:
# Look at a specific value from the int_rate column

df['int_rate'][3]

' 12.98%'

In [47]:
type(df['int_rate'][3])

str

Problems that we need to address with this column:

- String column that should be numeric
- Percent Sign `%` included with the number
- Leading space at the beginning of the string

However, we're not going to try and write exactly the right code to fix this column in one go. We're going to methodically build up to the code that will help us address these problems.


In [0]:
# Lets start with just fixing a single string.
# If we can fix one, we can usually fix all of them

int_rate = '  12.98%'

In [50]:
int_rate.strip()

'12.98%'

In [52]:
int_rate.strip("%")

'  12.98'

In [53]:
int_rate.strip().strip('%')

'12.98'

In [54]:
# "Casting" take a value of one datatype and change it to another datatype
# that is called "Casting"
# "Cast the string value to a float"

float('8.75')

8.75

In [55]:
# "Cast" the string value to a float
float(int_rate.strip().strip('%'))

12.98

In [56]:
type(float(int_rate.strip().strip('%')))

float

### Write a function to make our solution reusable!

In [0]:
# Write a function that can do what we have written above to any 
# string that is passsed to it.

def int_rate_to_float(cell_contents):
  return float(cell_contents.strip().strip('%'))

In [63]:
# Test out our function by calling it on our example
int_rate_to_float(int_rate)

12.98

In [64]:
# is the data type correct?
type(int_rate_to_float(int_rate))

float

### Apply our solution to every cell in a column

In [66]:
new_list = []

for cell in df['int_rate']:
  new_list.append(int_rate_to_float(cell))
# Look at the  first 10 values of the list
new_list[0:10]

[14.47, 10.33, 15.02, 12.98, 16.14, 12.98, 15.02, 16.14, 14.47, 13.56]

In [67]:
df['int_rate_cleaned'] = pd.Series(new_list)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3             Firefighter   10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     55000.0            Verified  Dec-2018  Charged Off          n   
1    100000.0        Not Verified  Dec-2018      Current          n   
2     90000.0            Verified  Dec-2018   Fully Paid          n   
3     95000.0     Source Verified  Dec-2018      Current          n   
4     56000.0     Source Verified  Dec-2018      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Jun-2006               0                     NaN   
1            0         Oct-1996               3                    39.0   
2            0         May-2007               1                     NaN   
3            0         Mar-2009               2                    73.0   
4            0         Dec-2004               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019           546

In [69]:
# What type of data is held in our new column?

# Look at the datatypes of the last 5 columns
df.dtypes[-5:]

settlement_date           object
settlement_amount        float64
settlement_percentage    float64
settlement_term          float64
int_rate_cleaned         float64
dtype: object

## Improve this code with the .appl()function


In [70]:
df['int_rate_cleaned_from_apply'] = df['int_rate'].apply(int_rate_to_float)

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3             Firefighter   10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status   issue_d  loan_status pymnt_plan  \
0     55000.0            Verified  Dec-2018  Charged Off          n   
1    100000.0        Not Verified  Dec-2018      Current          n   
2     90000.0            Verified  Dec-2018   Fully Paid          n   
3     95000.0     Source Verified  Dec-2018      Current          n   
4     56000.0     Source Verified  Dec-2018      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0         Jun-2006               0                     NaN   
1            0         Oct-1996               3                    39.0   
2            0         May-2007               1                     NaN   
3            0         Mar-2009               2                    73.0   
4            0         Dec-2004               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019           546

## Challenge

We can create a new column with our cleaned values or overwrite the original, whatever we think best suits our needs. On your assignment you will take the same approach in trying to methodically build up the complexity of your code until you have a few lines that will work for any cell in a column. At that point you'll contain all of that functionality in a reusable function block and then use the `.apply()` function to... well... apply those changes to an entire column.

# [Objective](#pandas-apply) Modify and Create Columns using `.apply()`



## Overview

We're already seen one example of using the `.apply()` function to clean up a column. Lets see if we can do it again, but this time on a slightly more complicated use case.

Remember, the goal here is to write a function that will work correctly on any **individual** cell of a specific column. Then we can reuse that function on those individual cells of a dataframe column via the `.apply()` function.

Lets clean up the emp_title "Employment Title" column!

## Follow Along

First we'll try and diagnose how bad the problem is and what improvements we might be able to make.

In [72]:
# Look at the top 20 employment titles
df['emp_title'].value_counts(dropna=False)[:20]

NaN                   20942
Teacher                2090
Manager                1773
Registered Nurse        951
Driver                  924
RN                      726
Supervisor              697
Sales                   580
Project Manager         526
General Manager         523
Office Manager          521
Owner                   420
Director                402
Operations Manager      387
Truck Driver            387
Nurse                   326
Engineer                325
Sales Manager           304
manager                 301
Supervisor              270
Name: emp_title, dtype: int64

In [73]:
# How many different unique employment titles are there currently?

df['emp_title'].value_counts()

Teacher                      2090
Manager                      1773
Registered Nurse              951
Driver                        924
RN                            726
                             ... 
Dental Office Admistrator       1
Senior Salesperson              1
Kitchen supervisor              1
Quota Management Analyst        1
Buzines partner                 1
Name: emp_title, Length: 43889, dtype: int64

In [74]:
len(df['emp_title'].value_counts())

43889

In [76]:
len(df['emp_title'].unique())

43890

In [78]:
# How often is the employment_title null?
df['emp_title'].isnull().sum()

20942

What are some possible reasons as to why a person's employment title may have not been provided?

In [0]:
# Create some examples that represent the cases that we want to clean up

examples = ['Owner', 'Supervisor', 'Project Manager', np.NaN]

In [90]:
# Write a function to clean up these use cases and increase uniformity.
df.clean_title(job_title)
if isinstance(jobs_title, str):
return job_title.title().strip()
else:
 return "unknown"

for title in examples:
  print(clean_title(title))

IndentationError: ignored

In [93]:
# list comprehensions can combine function calls and for loops over lists
# into one succinct and fairly readable single line of code.
[clean_title(title)] for title in examples]

SyntaxError: ignored

In [96]:
# We have a function that works as expected. Lets apply it to our column.
# This time we'll overwrite the original 

df['emp_title'] = df['emp_title'].apply(clean_title)

df.head()

NameError: ignored

We can use the same code as we did earlier to see how much progress was made.


In [99]:
# Look at the top 20 employment titles

df['emp_title'].value_counts(ascending=False)[:20]

Teacher                     2090
Manager                     1773
Registered Nurse             951
Driver                       924
RN                           726
Supervisor                   697
Sales                        580
Project Manager              526
General Manager              523
Office Manager               521
Owner                        420
Director                     402
Truck Driver                 387
Operations Manager           387
Nurse                        326
Engineer                     325
Sales Manager                304
manager                      301
Supervisor                   270
Administrative Assistant     269
Name: emp_title, dtype: int64

In [100]:
# How many different unique employment titles are there currently?

len(df['emp_title'].value_counts())

43889

In [101]:
# How often is the employment_title null (NaN)?
df['emp_title'].isnull().sum()

20942

## Challenge

Using the .apply() function isn't always about creating new columns on a dataframe, we can use it to clean up or modify existing columns as well. 

# [Objective](#dates-and-times) Work with Dates and Times with Pandas

## Overview

Pandas has its own datatype datatype that makes it extremely convenient to convert strings that are in standard date formates to datetime objects and then use those datetime objects to either create new features on a dataframe or work with the dataset in a timeseries fashion. 

This section will demonstrate how to take a column of date strings, convert it to a datetime object and then use the datetime formatting `.dt` to access specific parts of the date (year, month, day) to generate useful columns on a dataframe.

## Follow Along

### Work with Dates 

pandas documentation
- [to_datetime](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)
- [Time/Date Components](https://pandas.pydata.org/pandas-docs/stable/timeseries.html#time-date-components) "You can access these properties via the `.dt` accessor"

Many of the most useful date columns in this dataset have the suffix `_d` to indicate that they correspond to dates.

We'll use a list comprehension to print them out

In [106]:
[col for col in df col.endswith('_d')]

SyntaxError: ignored

Lets look at the string format of the `issue_d` column

In [104]:
df('issue_d')[:10]

TypeError: ignored

Because this string format %m-%y is a common datetime format, we can just let Pandas detect this format and translate it to the appropriate datetime object.

In [107]:
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)

df.dtypes[:15]

loan_amnt                       int64
funded_amnt                     int64
funded_amnt_inv               float64
term                           object
int_rate                       object
installment                   float64
grade                          object
sub_grade                      object
emp_title                      object
emp_length                     object
home_ownership                 object
annual_inc                    float64
verification_status            object
issue_d                datetime64[ns]
loan_status                    object
dtype: object

Now we can see that the `issue_d` column has been changed to hold `datetime` objects.

Lets look at one of the cells specifically to see what a datetime object looks like:

In [108]:
df['issue_d'].iloc[0]

Timestamp('2018-12-01 00:00:00')

You can see how the month and year have been indicated by the strings that were contained in the column previously, and that the rest of the values have been inferred.

In [109]:
df['issue_d'].head().values

array(['2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000', '2018-12-01T00:00:00.000000000',
       '2018-12-01T00:00:00.000000000'], dtype='datetime64[ns]')

We can use the `.dt` accessor to now grab specific parts of the datetime object. Lets grab just the year from the all of the cells in the `issue_d` column

In [114]:
df['issue_d'.dt.year

SyntaxError: ignored

Now the month.

In [115]:
df['emp_title'].dt.month

AttributeError: ignored

It's just that easy! Now, instead of printing them out, lets add these year and month values as new columns on our dataframe. Again, you'll have to scroll all the way over to the right in the table to see the new columns.

In [128]:
# look at the datatypes of our new columns
df.dtypes[-2:]

int_rate_cleaned               float64
int_rate_cleaned_from_apply    float64
dtype: object

Because all of these dates come from Q4 of 2018, the `issue_d` column isn't all that interesting. Lets look at the `earliest_cr_line` column, which is also a string, but that could be converted to datetime format.

We're going to create a new column called `days_from_earliest_credit_to_issue`

It's a long column header, but think about how valuable this piece of information could be. This number will essentially indicate the length of a person's credit history and if that is correlated with repayment or other factors could be a valuable predictor!

In [129]:
df['earliest_cr_line'].head()

0    Jun-2006
1    Oct-1996
2    May-2007
3    Mar-2009
4    Dec-2004
Name: earliest_cr_line, dtype: object

In [0]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], infer_datetime_format=True)

What we're about to do is so cool! Pandas' datetime format is so smart that we can simply use the subtraction operator `-` in order to calculate the amount of time between two dates. 

Think about everything that's going on under the hood in order to give us such straightforward syntax! Handling months of different lengths, leap years, etc. Pandas datetime objects are seriously powerful!

In [136]:
df['days_from_earliest_credit_to_issue'] = (df['issue_d'] - df['earliest_cr_line']).dt.days

df.head()

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  installment  \
0       6000         6000           6000.0   36 months   14.47%       206.44   
1      24000        24000          24000.0   36 months   10.33%       778.14   
2      30000        30000          30000.0   36 months   15.02%      1040.26   
3       8000         8000           8000.0   36 months   12.98%       269.48   
4      22400        22400          22400.0   60 months   16.14%       546.40   

  grade sub_grade                emp_title emp_length home_ownership  \
0     C        C2                  Manager  10+ years           RENT   
1     B        B1      Maintenance Manager  10+ years       MORTGAGE   
2     C        C3             Firefighter   10+ years           RENT   
3     B        B5               Controller    4 years       MORTGAGE   
4     C        C4  Paramedic/ Fire Fighter  10+ years            OWN   

   annual_inc verification_status    issue_d  loan_status pymnt_plan  \
0     55000.0            Verified 2018-12-01  Charged Off          n   
1    100000.0        Not Verified 2018-12-01      Current          n   
2     90000.0            Verified 2018-12-01   Fully Paid          n   
3     95000.0     Source Verified 2018-12-01      Current          n   
4     56000.0     Source Verified 2018-12-01      Current          n   

              purpose                    title zip_code addr_state    dti  \
0         credit_card  Credit card refinancing    104xx         NY  10.65   
1  debt_consolidation       Debt consolidation    606xx         IL  19.14   
2  debt_consolidation       Debt consolidation    681xx         NE  22.23   
3         credit_card  Credit card refinancing    480xx         MI   9.44   
4         credit_card  Credit card refinancing    234xx         VA  18.47   

   delinq_2yrs earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0            0       2006-06-01               0                     NaN   
1            0       1996-10-01               3                    39.0   
2            0       2007-05-01               1                     NaN   
3            0       2009-03-01               2                    73.0   
4            0       2004-12-01               0                    75.0   

   mths_since_last_record  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0                     NaN         6        0       4196      91.2%          8   
1                     NaN        14        0       8547      27.7%         28   
2                     NaN         8        0      25754        48%         23   
3                     NaN         7        0      11434      33.3%         20   
4                     NaN        15        0      13555      38.2%         22   

  initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
0                   w       0.00           0.00   1106.520000   
1                   w   16807.30       16807.30   9317.020000   
2                   w       0.00           0.00  30883.673712   
3                   w    1707.34        1707.34   6855.630000   
4                   w   19230.91       19230.91   6526.670000   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          1106.52           546.14         272.38                 0.0   
1          9317.02          7192.70        2124.32                 0.0   
2         30883.67         30000.00         883.67                 0.0   
3          6855.63          6292.66         562.97                 0.0   
4          6526.67          3169.09        3357.58                 0.0   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0       288.0                    51.84     Apr-2019           206.44   
1         0.0                     0.00     Dec-2019           778.14   
2         0.0                     0.00     Feb-2019         28382.51   
3         0.0                     0.00     Dec-2019           269.48   
4         0.0                     0.00     Dec-2019        

What's oldest credit history that was involved in Q4 2018? 

In [137]:
df['days_from_earliest_credit_to_issue'].describe()

count    128397.000000
mean       5859.918464
std        2886.617678
min        1126.000000
25%        4049.000000
50%        5266.000000
75%        7244.000000
max       25171.000000
Name: days_from_earliest_credit_to_issue, dtype: float64

25,171 days is ~ 68.96 years of credit history!

## Challenge

Pandas' datetime format is so easy to work with that there's really no excuse for not using dates to make features on a dataframe! Get ready to practice more of this on your assignment.